code.js

```diff
@@ -0,0 +1,68 @@
+async function pMergesort(array){
+  
+    let arrLen = array.length; 
+
+    // logic from mergesort repo
+    for (let i=1;i <=(arrLen-1); i = (2*i)){
+
+        //keeps track of the merge promises
+        let mergeProms = []; 
+        //Also logic from mergesort repo
+        for (let j = 0; j < (arrLen - 1); j = (j + (2 * i))){
+
+            let mid = (j + i - 1); 
+            let last = Math.min((j + (2 * i) - 1), (arrLen - 1));
+
+            // creaction of the promises that will merge the grouping
+            mergeProms.push(
+                new Promise(resolve =>{
+
+                    let left = j; 
+                    let right = (mid + 1); 
+
+                    // so we dont mess with the info other are working with
+                    let merged = [...array];
+
+                    // the actual merge logic, simular to mergesort repo
+                    while (left <= mid && right <= last){
+                        if(merged[left] > merged[right]){
+
+                            let tmp = merged[right];
+
+                            for (let k = right; k > left; k--){
+                                merged[k] = merged[k - 1];
+                            }
+                            merged[left] = tmp;
+                            left++;
+                            right++;
+                            mid++;
+                        } 
+                        else{
+                            left++; 
+                        }
+                    }
+
+                    // the resolution to a promise with the new grouping
+                    resolve({start: j, data: merged.slice(j, (last + 1)) });
+
+                })
+            );
+        }
+
+        // we wait for all the promises before moving on as to not mess anything up
+        let results = await Promise.all(mergeProms);
```

Comment:
But this relies on all promises being run in the order in which you push them, i.e. there can't be any parallelism?

Reply:
I think there would still be parallelism right? I thought that the order they were pushed didn't matter because the parallel part would be that the promises start working as they are created and then that work is done concurrently, or at least that was my understanding, and the result there is just waiting for the work to be finished before moving on.

Reply:
But the merge code assumes that the parts are sorted -- if they are run in an arbitrary order, that might not be the case.

Reply:
I thought that the pieces being merged would always be technically sorted, or it assumes the groups are already sorted, because when it recursed down the groups given to the next iteration would be sorted, so like it goes to individual elements and then merges them and that makes that group sorted and then the groups to be merge are sorted before the next merging, but the act of merging starts happening as soon as the new promise is created which is the parallel part. So the merges can be happening together which is why we wait for them all to complete but the order the start or complete doesn't matter because we will wait for them all and they will result in the proper groups before every iteration handles the next set of merges. I don't know if my thought process is making sense when I put it in words.